pallets/avn-transaction-payment/src/lib.rs

```diff
@@ -166,3 +173,85 @@ pub mod pallet {
+
+/// Implements the transaction payment for a pallet implementing the `Currency`
+/// trait (eg. the pallet_balances) using an unbalance handler (implementing
+/// `OnUnbalanced`).
+///
+/// The unbalance handler is given 2 unbalanceds in [`OnUnbalanced::on_unbalanceds`]: fee and
+/// then tip.
+pub struct AvnCurrencyAdapter<C, OU>(PhantomData<(C, OU)>);
+
+/// Default implementation for a Currency and an OnUnbalanced handler.
+///
+/// The unbalance handler is given 2 unbalanceds in [`OnUnbalanced::on_unbalanceds`]: fee and
+/// then tip.
+impl<T, C, OU> OnChargeTransaction<T> for AvnCurrencyAdapter<C, OU>
+where
+    T: Config + pallet::Config<Currency = C>,
+    C: Currency<<T as frame_system::Config>::AccountId>,
+    C::PositiveImbalance: Imbalance<
+        <C as Currency<<T as frame_system::Config>::AccountId>>::Balance,
+        Opposite = C::NegativeImbalance,
+    >,
+    C::NegativeImbalance: Imbalance<
+        <C as Currency<<T as frame_system::Config>::AccountId>>::Balance,
+        Opposite = C::PositiveImbalance,
+    >,
+    OU: OnUnbalanced<NegativeImbalanceOf<C, T>>,
+{
+    type LiquidityInfo = Option<NegativeImbalanceOf<C, T>>;
+    type Balance = <C as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+    /// Withdraw the predicted fee from the transaction origin.
+    ///
+    /// Note: The `fee` already includes the `tip`.
+    fn withdraw_fee(
+        who: &<T as frame_system::Config>::AccountId,
+        _call: &<T as frame_system::Config>::RuntimeCall,
+        _info: &DispatchInfoOf<<T as frame_system::Config>::RuntimeCall>,
+        fee: Self::Balance,
+        tip: Self::Balance,
+    ) -> Result<Self::LiquidityInfo, TransactionValidityError> {
+        return <CurrencyAdapter<C, OU> as OnChargeTransaction<T>>::withdraw_fee(
+            who, _call, _info, fee, tip,
+        )
+    }
+
+    /// Hand the fee and the tip over to the `[OnUnbalanced]` implementation.
+    /// Since the predicted fee might have been too high, parts of the fee may
+    /// be refunded.
+    ///
+    /// Note: The `corrected_fee` already includes the `tip`.
+    fn correct_and_deposit_fee(
+        who: &<T as frame_system::Config>::AccountId,
+        _dispatch_info: &DispatchInfoOf<<T as frame_system::Config>::RuntimeCall>,
+        _post_info: &PostDispatchInfoOf<<T as frame_system::Config>::RuntimeCall>,
+        corrected_fee: Self::Balance,
+        tip: Self::Balance,
+        already_withdrawn: Self::LiquidityInfo,
+    ) -> Result<(), TransactionValidityError> {
+        if let Some(paid) = already_withdrawn {
+            // Calculate how much refund we should return
+            let refund_amount = paid.peek().saturating_sub(corrected_fee);
+
+            // refund to the the account that paid the fees. If this fails, the
+            // account might have dropped below the existential balance. In
+            // that case we don't refund anything.
```

Comment:
Why don't we refund anything if the account is below the existential balance? Because it will have been reaped by then? But refunding it could help it stay above the existential deposit. I think I need an offline session to understand this comment.

Reply:
This comment is from Substrate's code from [here](https://github.com/paritytech/substrate/blob/5a4f49bd254fb00c2d902398471cf669a63b45d0/frame/transaction-payment/src/payment.rs#L64)